### PR TITLE
feat(libs/log)!: replace underlying logger (go-kit) w/ slog

### LINKED ISSUE #1278: logging: Migrate to slog
We should strongly consider migrating to using <https://pkg.go.dev/log/slog> for logging, since it's now provided by the Go standard library [since 1.21](https://tip.golang.org/doc/go1.21#slog).

Having to manage fewer dependencies would be great.

### PR DESCRIPTION
**Pull request by melekes**
Closes #1278 

Benchmarks: https://github.com/betterstack-community/go-logging-benchmarks

API changes:
- Removes `NewSyncWriter` - not needed anymore
- Removes `TestingLoggerWithColorFn` - hard to replicate (may be done in the future if needed)
- Removes `NewTMLoggerWithColorFn`- hard to replicate (may be done in the future if needed)

Notes:
- multiple module keys will all appear in the log, meaning With("module", "node").With("module", "crypto") will produce `module=node module=crypto`

### CODE REVIEW

**Comment by melekes** [comment]
We need to check if https://github.com/cometbft/cometbft/issues/2847 can be solved by simply turning the debug level off by default in `slog`.

**Review by alesforz — APPROVED** [review]
+1 for `slog` 😄 
It would be cool if, in the future, we could standardize log messages to use [LogAttrs](https://pkg.go.dev/log/slog#Logger.LogAttrs) instead of the utility functions (`Info`, `Debug`, etc.), as this would make logging even more efficient.

**Comment by cason** [comment]
There is an important change on how the logs look like with this change. I just did a simple test, by running:

Original:

```
internal/consensus $ go test -v -run TestStateTimestamp_ProposalNotMatch
I[2024-10-31|13:58:16.470] service start                                module=events msg="Starting EventBus service" impl=EventBus
I[2024-10-31|13:58:16.470] service start                                module=pubsub msg="Starting PubSub service" impl=PubSub
I[2024-10-31|13:58:16.470] Resetting proposal info                      module=consensus height=1 round=1 proposer=5E7F6D15970E7972245F0B7F67FFA580C3643C6C
D[2024-10-31|13:58:16.470] Entering new round                           module=consensus height=1 round=1 previous=1/0/RoundStepNewHeight proposer=5E7F6D15970E7972245F0B7F67FFA580C3643C6C
D[2024-10-31|13:58:16.470] Entering propose step                        module=consensus height=1 round=1 current=1/1/RoundStepNewRound
D[2024-10-31|13:58:16.470] Propose step; not our turn to propose        module=consensus height=1 round=1 proposer=5E7F6D15970E7972245F0B7F67FFA580C3643C6C
```

New:

```
internal/consensus $ go test -v -run TestStateTimestamp_ProposalNotMatch
Oct 31 14:58:06.484 INF service start module=events msg="Starting EventBus service" impl=EventBus
Oct 31 14:58:06.484 INF service start module=events module=pubsub msg="Starting PubSub service" impl=PubSub
Oct 31 14:58:06.484 INF Resetting proposal info module=consensus height=1 round=1 proposer=3A7D9559E3B63150E53AFFBC6C76372DCA834467
Oct 31 14:58:06.484 DBG Entering new round module=consensus height=1 round=1 previous=1/0/RoundStepNewHeight proposer=3A7D9559E3B63150E53AFFBC6C76372DCA834467
Oct 31 14:58:06.484 DBG Entering propose step module=consensus height=1 round=1 current=1/1/RoundStepNewRound
Oct 31 14:58:06.484 DBG Propose step; not our turn to propose module=consensus height=1 round=1 proposer=3A7D9559E3B63150E53AFFBC6C76372DCA834467
```

I am not mentioning colors (I actually do prefer slog's one), but how the messages are formatted and presented.

**Comment by andynog** [comment]
> 
> I am not mentioning colors (I actually do prefer slog's one), but how the messages are formatted and presented.
> 
> I do prefer the original timestamp and level (I/D) and I like the tab between message and details.

This might be breaking for operators (devops) if they use some log tools (e.g. Splunk), they'd need to change the logic on their side. A lot of them might use logs for alerting/observability.

**Comment by alesforz** [comment]
> > I am not mentioning colors (I actually do prefer slog's one), but how the messages are formatted and presented.
> > I do prefer the original timestamp and level (I/D) and I like the tab between message and details.
> 
> This might be breaking for operators (devops) if they use some log tools (e.g. Splunk), they'd need to change the logic on their side. A lot of them might use logs for alerting/observability.

🤔 good point. 
Since we are using a third-party handler to have colorized output (let's call it `colorHandler`), we then probably need to create a custom handler that wraps `colorHandler` to modify the timestamp's format, then calls `colorHandler` to render the message.

I can have a look how to do it, unless @melekes is already working on this?

**Comment by melekes** [comment]
> This might be breaking for operators (devops) if they use some log tools (e.g. Splunk), they'd need to change the logic on their side. A lot of them might use logs for alerting/observability.

Let's ask instead of just making assumptions. I've heard that many operators use JSON formatting precisely because they use log tools. Text output is mostly for humans afaik.

> I do prefer the original timestamp and level (I/D) and I like the tab between message and details.

kk. anyone else?

**Comment by melekes** [comment]
here's the patch for old format @cason

[mypatch.patch](https://github.com/user-attachments/files/17597077/mypatch.patch)

I personally like the new format 👍

**Comment by melekes** [comment]
Adopted Cosmos SDK `log.Logger` interface so that Cosmos SDK can drop their shim.

**Comment by cason** [comment]
> here's the patch for old format @cason
> 
> [mypatch.patch](https://github.com/user-attachments/files/17597077/mypatch.patch)
> 
> I personally like the new format 👍

You are going to kill me, but any change to make this configurable?

File: docs/tutorials/forum-application/6.main.md [hunk 1]
```diff
@@ -53,7 +53,7 @@ func main() {
 		log.Fatalf("failed to read config: %v", err)
 	}
 
-	logger := cmtlog.NewTMLogger(cmtlog.NewSyncWriter(os.Stdout))
+	logger := cmtlog.NewTMLogger(os.Stdout)
```
Comment (cason) [inline docs/tutorials/forum-application/6.main.md:56]:
```suggestion
	logger := cmtlog.NewLogger(os.Stdout)
```

File: docs/tutorials/forum-application/6.main.md [hunk 2]
```diff
@@ -140,7 +140,7 @@ the creation of the database, it logs the failure and exits.
 The program proceeds to create an instance of the ForumApp object using `forum.NewForumApp(dbPath, appConfigPath)`.
 If there is an error during the creation of the ForumApp instance, it logs the failure and exits.
 
-The program then sets up logging using `cmtlog.NewTMLogger(cmtlog.NewSyncWriter(os.Stdout))` and parses the log level
+The program then sets up logging using `cmtlog.NewTMLogger(os.Stdout)` and parses the log level
```
Comment (cason) [inline docs/tutorials/forum-application/6.main.md:143]:
```suggestion
The program then sets up logging using `cmtlog.NewLogger(os.Stdout)` and parses the log level
```

File: docs/tutorials/go-built-in.md [hunk 3]
```diff
@@ -654,7 +654,7 @@ func main() {
         log.Fatalf("failed to load node's key: %v", err)
     }
 
-    logger := cmtlog.NewTMLogger(cmtlog.NewSyncWriter(os.Stdout))
+    logger := cmtlog.NewTMLogger(os.Stdout)
```
Comment (cason) [inline docs/tutorials/go-built-in.md:657]:
```suggestion
    logger := cmtlog.NewLogger(os.Stdout)
```

File: docs/tutorials/go.md [hunk 4]
```diff
@@ -615,7 +615,7 @@ func main() {
     }()
 
     app := NewKVStoreApplication(db)
-    logger := cmtlog.NewTMLogger(cmtlog.NewSyncWriter(os.Stdout))
+    logger := cmtlog.NewTMLogger(os.Stdout)
```
Comment (cason) [inline docs/tutorials/go.md:618]:
```suggestion
    logger := cmtlog.NewLogger(os.Stdout)
```

File: libs/log/filter.go [hunk 5]
```diff
@@ -126,6 +134,10 @@ func (l *filter) With(keyvals ...any) Logger {
 	}
 }
 
+func (l *filter) Impl() any {
```
Comment (cason) [inline libs/log/filter.go:137]: Is this for SDK-compliance?

File: libs/log/filter_test.go
Comment (cason) [inline libs/log/filter_test.go:1]: Can we add some tests for the new WARN level?

File: libs/log/logger.go [hunk 6]
```diff
@@ -2,29 +2,115 @@ package log
...
+	// passed to calls to Info, Warn, Debug or Error.
 	With(keyvals ...any) Logger
+
+	// Impl returns the underlying logger implementation.
+	// It is used to access the full functionalities of the underlying logger.
+	// Advanced users can type cast the returned value to the actual logger.
+	Impl() any
+}
+
+type tmLogger struct {
```
Comment (cason) [inline libs/log/logger.go:31]: Can we avoid `tm` here?

File: libs/log/logger_test.go
Comment (cason) [inline libs/log/logger_test.go:1]: Again, should we add some tests for WARN level?

**Review by cason — APPROVED** [review]
Great job.

Two possible follow-ups:
- Formatting: could we render it configurable, can we get the same format used until now?
- Warn level: document, test, and of course use it.

**Comment by cason** [comment]
> I do prefer the original timestamp and level (I/D) and I like the tab between message and details.

I mean, this is not an aesthetic question for me. It is just the fact that I've got used to parse the log messages in the way they currently are. And I assume that a lot of people working with Comet (e.g., validators) have too.

File: libs/log/filter.go (hunk 5 above)
Comment (melekes) [reply libs/log/filter.go:137]: yes

**Comment by jmalicevic** [comment]
> > I do prefer the original timestamp and level (I/D) and I like the tab between message and details.
> 
> I mean, this is not an aesthetic question for me. It is just the fact that I've got used to parse the log messages in the way they currently are. And I assume that a lot of people working with Comet (e.g., validators) have too.

I prefer the padding too because I am used to looking only at the right side when debugging.

**Comment by melekes** [comment]
> > > I do prefer the original timestamp and level (I/D) and I like the tab between message and details.
> > 
> > 
> > I mean, this is not an aesthetic question for me. It is just the fact that I've got used to parse the log messages in the way they currently are. And I assume that a lot of people working with Comet (e.g., validators) have too.
> 
> I prefer the padding too because I am used to looking only at the right side when debugging.

alright 😄 I give up [`1e80922` (#4374)](https://github.com/cometbft/cometbft/pull/4374/commits/1e8092215bec2a10581d9613c8cb693c617d312b)

**Comment by cason** [comment]
The padding is still not working as before:

```
internal/consensus $ go test -v -run TestStateTimestamp_ProposalNotMatch
$ go test -v -run TestStateTimestamp_ProposalNotMatch                                                                                                                                     [abc-0020-control-character-injection] 
=== RUN   TestStateTimestamp_ProposalNotMatch
2024-11-04T15:06:04.417 INF service start			 module=events msg="Starting EventBus service" impl=EventBus
2024-11-04T15:06:04.417 INF service start			 module=events module=pubsub msg="Starting PubSub service" impl=PubSub
2024-11-04T15:06:04.417 INF Resetting proposal info			 module=consensus height=1 round=1 proposer=6AA1890A4E574B7A5E28C8BBC1C7B5D48693931F
2024-11-04T15:06:04.417 DBG Entering new round			 module=consensus height=1 round=1 previous=1/0/RoundStepNewHeight proposer=6AA1890A4E574B7A5E28C8BBC1C7B5D48693931F
2024-11-04T15:06:04.417 DBG Entering propose step			 module=consensus height=1 round=1 current=1/1/RoundStepNewRound
2024-11-04T15:06:04.417 DBG Propose step; not our turn to propose			 module=consensus height=1 round=1 proposer=6AA1890A4E574B7A5E28C8BBC1C7B5D48693931F
```

I think it is using TABs, which produce different padding depending on the length of the general (first) message.

**Comment by melekes** [comment]
```
2024-11-04T20:13:30.248 INF service start                                        module=events msg="Starting EventBus service" impl=EventBus
2024-11-04T20:13:30.248 INF service start                                        module=events module=pubsub msg="Starting PubSub service" impl=PubSub
2024-11-04T20:13:30.248 INF Resetting proposal info                              module=consensus height=1 round=1 proposer=77DF2FD7540AA36F2E152F22B2CC583A76A359D0
2024-11-04T20:13:30.248 DBG Entering new round                           module=consensus height=1 round=1 previous=1/0/RoundStepNewHeight proposer=77DF2FD7540AA36F2E152F22B2CC583A76A359D0
2024-11-04T20:13:30.248 DBG Entering propose step                                module=consensus height=1 round=1 current=1/1/RoundStepNewRound
2024-11-04T20:13:30.248 DBG Propose step; not our turn to propose                module=consensus height=1 round=1 proposer=77DF2FD7540AA36F2E152F22B2CC583A76A359D0
2024-11-04T20:13:30.248 INF service start                                        module=consensus msg="Starting TimeoutTicker service" impl=TimeoutTicker
2024-11-04T20:13:30.248 DBG Starting timeout routine                             module=consensus
2024-11-04T20:13:30.248 DBG Received tick                                        module=consensus old_ti="{Duration:0s Height:0 Round:0 Step:RoundStepUnknown}" new_ti="{Duration:41ms Height:1 Round:1 Step:RoundStepPropose}"
```

there's still a one-off error somewhere, but can't fix it.

File: libs/log/logger.go [hunk 7]
```diff
@@ -1,30 +1,155 @@
...
+	return &baseLogger{logger}
+}
+
+// tabHandler is a slog.Handler that adds two tabs between the message and the attributes.
+type tabHandler struct {
+	h slog.Handler
+}
+
+func (th tabHandler) Handle(ctx context.Context, r slog.Record) error {
+	formattedMsg := r.Message
```
Comment (cason) [inline libs/log/logger.go:128]:
```suggestion
	formattedMsg := fmt.Sprintf("%-44s", r.Message)
```

File: libs/log/logger.go [hunk 8]
```diff
@@ -1,30 +1,155 @@
...
+
+	maxMsgLen := 40
```
Comment (cason) [inline libs/log/logger.go:130]: And you can remove all this logic.

**Review by cason — COMMENTED** [review]
Suggestion on how to get the same format as before, in terms of padding.

File: libs/log/logger.go (hunk 8 above)
Comment (melekes) [reply libs/log/logger.go]: thanks

**Pull request merged**